Add Aws Lambda execution context post (#61)
ippontech/blog-usa: posts/lambda-execution-context-demystified.md

---
authors:
- Raphael Brugier
tags:
- AWS
- AWS Lambda
- Java
date: 2018-10-17T18:05:14.000Z
title: "AWS Lambda Execution context in Java demystified"

> image: image: https://raw.githubusercontent.com/ippontech/blog-usa/master/images/2017/12/te-le-chargement.png

Reviewer: ```suggestion
image: https://raw.githubusercontent.com/ippontech/blog-usa/master/images/2017/12/te-le-chargement.png
```

---

In a [previous article](https://blog.ippon.tech/serverless-architectures/), we described how Serverless architectures enable scalability and can reduce costs. As of now, the most popular cloud provider for Serverless functions is still AWS with its Lambda service.

In this article, I will attempt to demystify the _Execution Context_ of a Lambda invocation and how you can take advantage of it. More specifically for Lambdas written in Java, but this can apply to any language.

# Lambda Execution Context

In the [AWS Lambda documentation](https://docs.aws.amazon.com/lambda/latest/dg/running-lambda-code.html), AWS describes an Execution Context as a “temporary runtime environment that initializes any external dependencies of your Lambda”.

> The _Execution Context_ is the invisible stack that AWS creates for your Lambda in order to execute the function. AWS maintains the _Context_ ready to accept new invocations of the function for an unknown amount of time[^1] and then delete it to free up some resource. Between each invocation, AWS freezes and unfreezes the Context.

Reviewer: ```suggestion
The _Execution Context_ is the invisible stack that AWS creates for your Lambda in order to execute the function. AWS maintains the _Context_ ready to accept new invocations of the function for an unknown amount of time[^1] and then deletes it to free up some resource. Between each invocation, AWS freezes and unfreezes the Context.
```

For a Lambda written in Java, this is how I like to represent the different layers of the _Execution Context_:

![](https://raw.githubusercontent.com/ippontech/blog-usa/master/images/2018/10/lambda-java-execution-context.png)

- An EC2 instance runs the entire stack. It’s not possible to physically access the instance, and AWS does not share information about this layer
- Container: a container is started to run each Lambda in isolation. Multiple containers run in the same instance
- JVM: For Lambdas written in Java, a Java Virtual Machine is started. AWS uses OpenJDK 8
- An instance of the JavaHandler class where the function is defined
- The function itself

Between each invocation of the `handleRequest` method, the same _Execution Context_ is reused. This, at least until the Lambda function does not receive requests anymore and AWS destroys the _Context_ to reclaim the resources.

# Why it matters

> As you can imagine, starting-up the _Execution Context_ takes more time than just invoking the function. 

Reviewer: "takes more time than just invoking the function" -> not necessarily true / depends on the function

Author: ah yes, absolutely. Especially now that a Lambda can run up to 15 min :) 
```suggestion
As you can imagine, starting-up the _Execution Context_ can take some time but is inevitable before running the function for the first time.
```

        headers.put("X-Custom-Header", "application/json");
        return new GatewayResponse(output, headers, 200);
    }
}

```

1. The static block is executed when the _Execution Context_ loads the LambdaHandler class in the JVM for the first time. The random value is stored in a static field.
2. The constructor is executed when the _Execution Context_ instantiates a LambdaHandler instance for the first time. The random value is stored in a field of the instance.
3. The handleRequest method is executed every time the Lambda function is invoked and the random value stored in a local variable.

Accessing the endpoint multiple times returns the following message:
```json
{ "1 - static value": "0.936418", " 2 - constructor value": "0.400978", "3 - invocation value": "0.362746"}
{ "1 - static value": "0.936418", " 2 - constructor value": "0.400978", "3 - invocation value": "0.969463"}
{ "1 - static value": "0.936418", " 2 - constructor value": "0.400978", "3 - invocation value": "0.0.723602"}
```

As expected, only the _invocation value_ changes between each invocation. The values generated by both the static block and the constructor remain the same.

> If you access the [endpoint](https://jz2ckwdte0.execute-api.us-east-1.amazonaws.com/Prod/execute/) from your browser multiple times, you will notice the same behavior where only the _invocation value_ changes between your requests. The static and constructor values remains the same. The values will probably be different than the example above, that is because AWS has destroyed the _Execution Context_ and a new one had to get created during your first request.

Reviewer: ```suggestion
If you access the [endpoint](https://jz2ckwdte0.execute-api.us-east-1.amazonaws.com/Prod/execute/) from your browser multiple times, you will notice the same behavior where only the _invocation value_ changes between your requests. The static and constructor values remain the same. The values will probably be different than the example above, that is because AWS has destroyed the _Execution Context_ and a new one had to get created during your first request.
```

> >Note: To deploy this Lambda in your own account, clone the [project repository](https://github.com/raphaelbrugier/aws-lambda-java-exec-context) and follow the instructions in the deploy section of the README file.

Reviewer: ```suggestion
>Note: To deploy this Lambda to your own account, clone the [project repository](https://github.com/raphaelbrugier/aws-lambda-java-exec-context) and follow the instructions in the deploy section of the README file.
```
